add: per request statement timeout using prefer header

taimoorzaeem opened the pull request:
Adds a feature to set `statement_timeout` using `Prefer: timeout` header. This also introduces a `PGRST129` error which is returned when the timeout preferred exceeds the per-role timeout, which prevents misuse of this feature.

Closes #4381.

- [ ] Implementation
- [ ] Tests
- [ ] Docs

steve-chavez commented on docs/references/api/preferences.rst:
```diff
@@ -296,3 +297,38 @@ With :ref:`RPC <functions>`, the preference is honored completely on the basis o
+
+You can set `statement_timeout <https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-STATEMENT-TIMEOUT>`_ for the request using this preference. This works in combination with ``handling=strict`` preference in the same header.
+
+Currently, the header only accepts integer value indicating the ``seconds`` that are set as timeout value. To demonstrate, see the following example:
```
I don't think this is going to change:
```suggestion
The header only accepts integer value indicating the ``seconds`` that are set as timeout value. To demonstrate, see the following example:
```

steve-chavez commented:
We should also fail in case a value exceeds the [per role timeout](https://docs.postgrest.org/en/v13/references/transactions.html#impersonated-role-settings). Otherwise it's abusable.

This failure should happen at the plan level, before hitting the db. I think it will need a new error code.

steve-chavez commented on src/PostgREST/Plan.hs:
```diff
@@ -144,29 +146,56 @@ data InfoPlan
+      -- It is safe to use fromJust because the string being read is guaranteed
+      -- to be number after striping the unit.
+      stripUnitAndGetInt unit val = fromJust $ BS.stripSuffix unit val >>= readMaybe
+      rtValueToSeconds :: ByteString -> Int
```
@taimoorzaeem Hm, so we can't avoid this logic with the [above snippet](https://github.com/PostgREST/postgrest/pull/4584/changes#r2679654379)?

taimoorzaeem replied:
Yes. If we try to avoid this logic using the _above snippet_, we would end with maintaining more state (keep extra pre-computed values, even when they are never used) so I think we should avoid it.

steve-chavez replied:
Sorry for late reply here.

> If we try to avoid this logic using the above snippet, we would end up with maintaining more state (keep extra pre-computed values, even when they are never used) so I think we should avoid it.

IMO we should cache the pg transformed value. 

- Having to repeat this transformation logic for each request at the Plan level would be inefficient.
- It's not much more memory considering we only take child roles of the connection role:
https://github.com/PostgREST/postgrest/blob/ffa393850a461ed02a0906833a52efe5fa844d43/src/PostgREST/Config/Database.hs#L145
- Also later on we could avoid this caching if `Prefer: timeout` is not used with https://github.com/PostgREST/postgrest/issues/2987.
- There is more risk for errors with this Haskell logic.. say if pg changes format of the statement_timeout, very unlikely but overall it feels safer to have the transformation at the db level.

steve-chavez commented on docs/references/api/preferences.rst:
```diff
@@ -296,3 +297,62 @@ With :ref:`RPC <functions>`, the preference is honored completely on the basis o
+
+.. code-block:: http
+
+  HTTP/1.1 500 Internal Server Error
```
I was thinking about mapping this error to 503 but this isn't correct as 503 indicates temporary state on which a later retry can succeed; this is not true for example when an index is missing on a column, all retries would fail the same. So 500 is the most apt code here.

That also reminds me of https://github.com/PostgREST/postgrest/issues/2967#issuecomment-2713224061, we should put `statement_timeout` on the Proxy-Status.

steve-chavez commented on docs/references/api/preferences.rst:
```diff
@@ -296,3 +297,62 @@ With :ref:`RPC <functions>`, the preference is honored completely on the basis o
+
+  {
+      "code": "PGRST129",
+      "message": "Timeout preference value cannot exceed statement_timeout of role",
```
Maybe the message should give less details:
```suggestion
      "message": "Timeout preference exceeded maximum allowed",
```

steve-chavez commented on docs/references/api/preferences.rst:
```diff
@@ -296,3 +297,62 @@ With :ref:`RPC <functions>`, the preference is honored completely on the basis o
+
+You can set `statement_timeout <https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-STATEMENT-TIMEOUT>`_ for the request using this preference. This works in combination with ``handling=strict`` preference in the same header.
+
+The header only accepts integer value indicating the ``seconds`` that are set as timeout value. To demonstrate, see the following example:
```
Should we indicate here that `timeout=0` is a noop?

steve-chavez commented on src/PostgREST/ApiRequest/Preferences.hs:
```diff
@@ -171,8 +182,9 @@ fromHeaders allowTxDbOverride acceptedTzNames headers =
     prefMap :: ToHeaderValue a => [a] -> Map.Map ByteString a
     prefMap = Map.fromList . fmap (\pref -> (toHeaderValue pref, pref))
 
+
```
```suggestion
```

steve-chavez commented on src/PostgREST/Config/Database.hs:
```diff
@@ -155,33 +157,46 @@ queryRoleSettings pgVer prepared =
+      role_timeout_setting AS (
+        SELECT rolname, extract(epoch from value::interval)::int as value -- transform to seconds
+        FROM kv_settings
+        WHERE key = 'statement_timeout'
```
Forgot about this, if there's a global `statement_timeout` (this can be set at the cluster level or at the db level) but no role timeout, it would be bad if we allowed clients to surpass this global too. So we would need to refine the query to obtain the global timeout as implicit default for the role.

steve-chavez commented on test/io/test_prefer_header.py:
```diff
@@ -0,0 +1,150 @@
+        # set global statement_timeout to 3s using postgres superuser
+        os.system(
+            f'psql -d {env["PGDATABASE"]} -U postgres -h {env["PGHOST"]} --set ON_ERROR_STOP=1 -a -c "ALTER DATABASE postgres SET statement_timeout TO \'3s\';"'
+        )
```
Nice, you nailed it 💯

steve-chavez commented on test/io/fixtures/roles.sql:
```diff
@@ -1,18 +1,23 @@
 CREATE ROLE postgrest_test_w_superuser_settings;
+CREATE ROLE postgrest_test_timeout_ms;
+CREATE ROLE postgrest_test_timeout_s;
+CREATE ROLE postgrest_test_timeout_global; -- no timeout for this role, default to global (db/cluster) statement_timeout
```
Maybe a more descriptive name? Should help with readability on the tests.
```suggestion
CREATE ROLE postgrest_test_no_timeout; -- no timeout for this role, default to global (db/cluster) statement_timeout
```

steve-chavez commented on src/PostgREST/Config/Database.hs:
```diff
@@ -173,8 +176,8 @@ queryRoleSettings pgVer prepared =
-        rowsWRoleSettings = [ (x, z) | (x, _, z) <- rs ]
-        rowsWIsolation    = [ (x, y) | (x, Just y, _) <- rs ]
+        rowsWRoleSettings = [ (x, z)  | (x, _, z) <- rs ]
+        rowsWIsolation    = [ (x, y)  | (x, Just y, _) <- rs ]
```
We better not change unrelated lines, helps for having a cleaner git history.
```suggestion
        rowsWRoleSettings = [ (x, z) | (x, _, z) <- rs ]
        rowsWIsolation    = [ (x, y) | (x, Just y, _) <- rs ]
```

steve-chavez commented on src/PostgREST/Query/PreQuery.hs:
```diff
@@ -50,7 +52,12 @@ txVarQuery dbActPlan AppConfig{..} AuthResult{..} ApiRequest{..} =
-    timezoneSql = maybe mempty (\(PreferTimezone tz) -> [setConfigWithConstantName ("timezone", tz)]) $ preferTimezone iPreferences
+    timezoneSql = maybe mempty (\(PreferTimezone tz) -> [setConfigWithConstantName ("timezone", tz)]) preferTimezone
+    timeoutSql = maybe mempty (\(PreferTimeout t) ->
+      if preferHandling == Just Strict && t > 0 -- 0 and -ve numbers are meaningless
```
Would 0 and negatives reach this module? They would be rejected at the Plan level before they reach Query right?

steve-chavez commented on src/PostgREST/Config/Database.hs:
```diff
@@ -183,6 +186,59 @@ queryRoleSettings pgVer prepared =
+      -- +-----------------------------------------------------------+
+      -- |   role1  |     0s             |             0             |
+      -- +-----------------------------------------------------------+
+      -- |   role2  |     300ms          |             0             |
```
Hm, this part confuses me because it should be 0.3 in seconds. Maybe say it's truncated?

steve-chavez commented on src/PostgREST/Config/Database.hs:
```diff
@@ -183,6 +186,59 @@ queryRoleSettings pgVer prepared =
+      --
+      -- However, both get cached to the same value. So, to differentiate
+      -- these two distinct cases, the query caches "disabled" timeout as NULL
+      -- and < 1s timeouts as 0.
```
Which line of code does this?

taimoorzaeem replied:
It's this one with the `nullif()` function:

```hs
-- src/PostgREST/Config/Database.hs:241:
      SELECT
        rolname,
        nullif(extract(epoch FROM statement_timeout::interval)::numeric, 0)::int as statement_timeout_in_seconds
      FROM role_timeouts
```

The `< 1s` automatically gets truncated to 0 because of `::int` cast.

taimoorzaeem committed:
add: per request statement timeout using prefer header

Adds a feature to set `statement_timeout` using `Prefer: timeout`
header. This also introduces a `PGRST129` error which is returned
when the timeout preferred exceeds the per-role or global timeout.

Signed-off-by: Taimoor Zaeem <taimoorzaeem@gmail.com>

wolfgangwalther commented on src/PostgREST/Plan.hs:
```diff
@@ -144,29 +145,42 @@ data InfoPlan
+failPreferTimeout Preferences{preferTimeout=(Just (PreferTimeout t)), preferHandling=Just Strict} AppConfig{..} AuthResult{..} =
+  case roleTimeout of
+    Nothing -> Right ()
+    Just rt -> when (t > rt) $ Left $ ApiRequestError $ TimeoutConstraintError rt
```
> Who would want to set the timeout header to a value less than the role setting? What for?

In way this would be similar to soft and hard limits in Linux, set via `ulimit`. You can vary the soft limit on your own as long as you stay below the hard limit. Just because there is some higher limit available doesn't mean that you will always make full use of it.

I think it only really becomes useful when you have both:
- a low default for the soft limit, which you can override per request, *and*
- a higher hard limit

IIUC, we currently only have the limit defaulting to the hard limit, which then becomes the odd case of "you can only lower your limit".

Edit: This specifically only answers this question - I'm not saying we should implement any of this and also made no judgement on the PR or the related issue.

mkleczek replied:
> > Who would want to set the timeout header to a value less than the role setting? What for?
> 
> In way this would be similar to soft and hard limits in Linux, set via `ulimit`. 

To be honest, I always had a problem with understanding the purpose of soft limits. The only sensible use case IMHO is "grace periods" - ie. exceeding soft limit triggers a warning and requires a corrective action in a specified time period. But that really only makes sense for disk quotas etc. (and is not exactly the use case of user-provided limit).

Our situation is: a user has a query to run. The query will take x seconds to execute or will fail if the limit is lower. As a user I want to make sure my query does not fail so the only sensible strategy (as I don't know how long it will take) is not to set any limit by myself (and deal with possibility that system enforced limit is lower).

Of course - we could implement tracking of resource usage (per user? per tenant?) and have a policy like `allow 10 queries above soft limit per month` - but I don't think we want to do that :)
